File: Cisco/GitHubList/Python/6DecodeString.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []

        for ch in s:
            if ch != "]":
                stack.append(ch)
            else:
                cur = []
                while stack and stack[-1] != "[":
                    cur.append(stack.pop())
                stack.pop()  # Remove '['

                num = []
                while stack and stack[-1].isdigit():
                    num.append(stack.pop())
                mul = int("".join(num[::-1]))

                stack.extend(cur[::-1] * mul)
        return "".join(stack)

File: Cisco/GitHubList/C++/6DecodeString.cpp
#include <string>
#include <stack>

using namespace std;

class Solution
{
public:
    string decodeString(string s)
    {
        stack<int> multiplierStack;
        stack<string> substringStack;
        string currSubstr;
        int currMultiplier;

        for (const auto &ch : s)
            if (isalpha(ch))
                currSubstr += ch;
            else if (isdigit(ch))
                currMultiplier = currMultiplier * 10 + (ch - '0');
            else if (ch == '[')
            {
                multiplierStack.push(currMultiplier), currMultiplier = 0;
                substringStack.push(currSubstr), currSubstr = "";
            }
            else // ch == ']'
            {
                string temp = currSubstr;
                currSubstr = substringStack.top(), substringStack.pop();
                currMultiplier = multiplierStack.top(), multiplierStack.pop();

                while (currMultiplier)
                {
                    currSubstr += temp;
                    --currMultiplier;
                }
            }

        return currSubstr;
    }
};